cdap-app-fabric/src/main/java/io/cdap/cdap/security/impersonation/DefaultUGIProvider.java

```diff
@@ -172,4 +200,52 @@ private File localizeKeytab(Location keytabLocation) throws IOException {
 
     return localKeytabFile.toFile();
   }
+
+  private Map<String, String> getRuntimeImpersonationProperties(NamespacedEntityId programId) {
+      if (programId instanceof ProgramRunId) {
+          ProgramRunId runId = ((ProgramRunId) programId);
+          RunRecordMeta runRecord = null;
+
+          long sleepDelayMs = TimeUnit.MILLISECONDS.toMillis(100);
+          long startTime = System.currentTimeMillis();
+          long timeoutMs = TimeUnit.SECONDS.toMillis(10);
+          while (System.currentTimeMillis() - startTime < timeoutMs) {
```

Comment:
Use the `Retries` and `RetryStrategy` class to simplify the logic.

Reply:
Done.